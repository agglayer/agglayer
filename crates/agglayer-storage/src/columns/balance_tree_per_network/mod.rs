use super::{ColumnSchema, BALANCE_TREE_PER_NETWORK_CF};

#[cfg(test)]
mod tests;

/// Column family for the balance tree per network.
///
/// ## Column definition
///
/// | key                               | value                                   |
/// | ---                               | --                                      |
/// | (`NetworkId`, `SmtKeyType::Root`) | (`hash(root.left)`, `hash(root.right)`) |
/// | (`NetworkId`, `hash(node)`)       | (`hash(node.left)`, `hash(node.right)`) |
/// | (`NetworkId`, `hash(leaf)`)       | (`hash(leaf)`)                          |
pub struct BalanceTreePerNetworkColumn;

impl ColumnSchema for BalanceTreePerNetworkColumn {
    type Key = crate::types::SmtKey;
    type Value = crate::types::SmtValue;

    const COLUMN_FAMILY_NAME: &'static str = BALANCE_TREE_PER_NETWORK_CF;
}
