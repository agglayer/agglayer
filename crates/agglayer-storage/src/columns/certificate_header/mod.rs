use agglayer_types::CertificateHeader;

use super::{ColumnSchema, CERTIFICATE_HEADER_CF};

#[cfg(test)]
mod tests;

/// Column family for the certificate headers.
///
/// ## Column definition
///
/// | key               | value                |
/// | --                | --                   |
/// | `CertificateId`   |  `CertificateHeader` |
pub struct CertificateHeaderColumn;

pub type Key = agglayer_types::CertificateId;
pub type Value = CertificateHeader;

impl ColumnSchema for CertificateHeaderColumn {
    type Key = Key;
    type Value = Value;

    const COLUMN_FAMILY_NAME: &'static str = CERTIFICATE_HEADER_CF;
}
