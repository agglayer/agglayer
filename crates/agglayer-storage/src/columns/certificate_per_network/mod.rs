use agglayer_types::CertificateId;
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, CERTIFICATE_PER_NETWORK_CF};

#[cfg(test)]
mod tests;

/// Column family for the certificate per network per height.
///
/// ## Column definition
///
/// | key                     | value           |
/// | --                      | --              |
/// | (`NetworkId`, `Height`) | `CertificateId` |
pub struct CertificatePerNetworkColumn;

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Key {
    pub(crate) network_id: u32,
    pub(crate) height: u64,
}

impl Codec for Key {}

impl ColumnSchema for CertificatePerNetworkColumn {
    type Key = Key;
    type Value = CertificateId;

    const COLUMN_FAMILY_NAME: &'static str = CERTIFICATE_PER_NETWORK_CF;
}
