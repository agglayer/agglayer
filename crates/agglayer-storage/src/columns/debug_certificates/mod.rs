use agglayer_types::{Certificate, CertificateId};

use super::{ColumnSchema, DEBUG_CERTIFICATES_CF};

/// Column family containing the certificates received.
///
/// ## Column definition
///
/// | key             | value           |
/// | --              | --              |
/// | `CertificateId` | `Certificate`   |
pub(crate) struct DebugCertificatesColumn;

impl ColumnSchema for DebugCertificatesColumn {
    type Key = CertificateId;
    type Value = Certificate;

    const COLUMN_FAMILY_NAME: &'static str = DEBUG_CERTIFICATES_CF;
}
