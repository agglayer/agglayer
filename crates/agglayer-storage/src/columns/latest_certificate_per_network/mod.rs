use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF};
use crate::types::NetworkId;

#[cfg(test)]
mod tests;

/// Column family for the latest settled certificate per network.
/// The key is the network_id and the value is the certificateID,
/// the height and the epoch_number.
///
/// ## Column definition
///
/// | key         | value                                |
/// | --          | --                                   |
/// | `NetworkId` | (`Proof`, `CertificateId`, `Height`) |
pub struct LatestSettledCertificatePerNetworkColumn;

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProvenCertificate(pub [u8; 32], pub u64, pub u64);

pub type Key = NetworkId;

impl Codec for ProvenCertificate {}

impl ColumnSchema for LatestSettledCertificatePerNetworkColumn {
    type Key = Key;
    type Value = ProvenCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF;
}
