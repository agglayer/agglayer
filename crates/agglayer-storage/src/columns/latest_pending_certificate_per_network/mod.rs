use agglayer_types::{CertificateId, Height, NetworkId};
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF};

/// Column family for the latest pending certificate per network.
/// The key is the network_id and the value is the certificateID and
/// the height.
///
/// ## Column definition
///
/// | key         | value                       |
/// | --          | --                          |
/// | `NetworkId` | (`CertificateId`, `Height`) |
pub struct LatestPendingCertificatePerNetworkColumn;

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PendingCertificate(pub CertificateId, pub Height);

pub type Key = NetworkId;

impl Codec for PendingCertificate {}

impl ColumnSchema for LatestPendingCertificatePerNetworkColumn {
    type Key = Key;
    type Value = PendingCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF;
}
