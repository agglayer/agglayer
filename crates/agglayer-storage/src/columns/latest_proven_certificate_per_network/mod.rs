use agglayer_types::{CertificateId, Height, NetworkId};
use serde::{Deserialize, Serialize};

use super::{Codec, ColumnSchema, LATEST_PROVEN_CERTIFICATE_PER_NETWORK_CF};

/// Column family for the latest proven certificate per network.
/// The key is the network_id and the value is the certificateID and
/// the height.
///
/// ## Column definition
///
/// | key         | value                       |
/// | --          | --                          |
/// | `NetworkId` | (`CertificateId`, `Height`) |
pub struct LatestProvenCertificatePerNetworkColumn;

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProvenCertificate(pub CertificateId, pub NetworkId, pub Height);

pub type Key = NetworkId;

impl Codec for ProvenCertificate {}

impl ColumnSchema for LatestProvenCertificatePerNetworkColumn {
    type Key = Key;
    type Value = ProvenCertificate;

    const COLUMN_FAMILY_NAME: &'static str = LATEST_PROVEN_CERTIFICATE_PER_NETWORK_CF;
}
