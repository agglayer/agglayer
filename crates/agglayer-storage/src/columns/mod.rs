use bincode::{DefaultOptions, Options};
use serde::{de::DeserializeOwned, Serialize};

use crate::error::Error;

pub fn default_bincode_options() -> impl bincode::Options {
    DefaultOptions::new()
        .with_big_endian()
        .with_fixint_encoding()
}

// State related CFs
pub const CERTIFICATE_PER_NETWORK_CF: &str = "certificate_per_network_cf";
pub const NULLIFIER_TREE_PER_NETWORK_CF: &str = "nullifier_tree_per_network_cf";
pub const BALANCE_TREE_PER_NETWORK_CF: &str = "balance_tree_per_network_cf";
pub const LOCAL_EXIT_TREE_PER_NETWORK_CF: &str = "local_exit_tree_per_network_cf";

// Metadata CFs
pub const CERTIFICATE_HEADER_CF: &str = "certificate_header_cf";
pub const LATEST_PROVEN_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_proven_certificate_per_network_cf";
pub const LATEST_SETTLED_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_settled_certificate_per_network_cf";
pub const LATEST_PENDING_CERTIFICATE_PER_NETWORK_CF: &str =
    "latest_pending_certificate_per_network_cf";
pub const METADATA_CF: &str = "metadata_cf";

// epochs related CFs
pub const PER_EPOCH_CERTIFICATES_CF: &str = "per_epoch_certificates_cf";
pub const PER_EPOCH_METADATA_CF: &str = "per_epoch_metadata_cf";
pub const PER_EPOCH_PROOFS_CF: &str = "per_epoch_proofs_cf";
pub const PER_EPOCH_END_CHECKPOINT_CF: &str = "per_epoch_end_checkpoint_cf";
pub const PER_EPOCH_START_CHECKPOINT_CF: &str = "per_epoch_start_checkpoint_cf";
pub const PER_EPOCH_TRANSACTION_HASH_PER_CERTIFICATE_INDEX: &str =
    "per_epoch_transaction_hash_per_certificate_index";

// Pending related CFs
pub const PENDING_QUEUE_CF: &str = "pending_queue_cf";
pub const PROOF_PER_CERTIFICATE_CF: &str = "proof_per_certificate_cf";

// debug CFs
pub const DEBUG_CERTIFICATES_CF: &str = "debug_certificates";

pub trait Codec: Sized + Serialize + DeserializeOwned {
    fn encode(&self) -> Result<Vec<u8>, Error> {
        Ok(default_bincode_options().serialize(self)?)
    }

    fn decode(buf: &[u8]) -> Result<Self, Error> {
        Ok(default_bincode_options().deserialize(buf)?)
    }
}

pub trait ColumnSchema {
    type Key: Codec;
    type Value: Codec;

    const COLUMN_FAMILY_NAME: &'static str;
}

// State
pub(crate) mod balance_tree_per_network;
pub(crate) mod certificate_per_network;
pub(crate) mod local_exit_tree_per_network;
pub(crate) mod nullifier_tree_per_network;

// Pending
pub(crate) mod pending_queue;
pub(crate) mod proof_per_certificate;

// Metadata
pub(crate) mod certificate_header;
pub mod latest_pending_certificate_per_network;
pub mod latest_proven_certificate_per_network;
pub mod latest_settled_certificate_per_network;
pub(crate) mod metadata;

// Debug
pub(crate) mod debug_certificates;

// PerEpoch
pub mod epochs {
    pub(crate) mod certificates;
    pub mod end_checkpoint;
    pub(crate) mod metadata;
    pub(crate) mod proofs;
    pub(crate) mod start_checkpoint;
    mod transaction_hash_per_certificate_index;
}
