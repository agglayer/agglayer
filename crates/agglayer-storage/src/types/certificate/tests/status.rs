use agglayer_types::{CertificateStatus, CertificateStatusError, Digest, NetworkId};
use alloy_primitives::Bytes;
use pessimistic_proof::{error::ProofVerificationError, ProofError};

use crate::columns::bincode_codec;

type Cse = CertificateStatusError;
type Pve = ProofVerificationError;

fn err(error: CertificateStatusError) -> CertificateStatus {
    CertificateStatus::InError { error }
}

fn proof_err(source: ProofError) -> CertificateStatus {
    err(CertificateStatusError::ProofGenerationError {
        generation_type: agglayer_types::GenerationType::Prover,
        source,
    })
}

fn ver_err(error: ProofVerificationError) -> CertificateStatus {
    err(CertificateStatusError::ProofVerificationFailed(error))
}

#[rstest::rstest]
#[case("pending", CertificateStatus::Pending)]
#[case("candidate", CertificateStatus::Candidate)]
#[case("proven", CertificateStatus::Proven)]
#[case("settled", CertificateStatus::Settled)]
#[case("err-prf-ho", proof_err(ProofError::HeightOverflow))]
#[case("err-prf-inp", proof_err(ProofError::InvalidNullifierPath))]
#[case("err-prf-is", proof_err(ProofError::InvalidSignature))]
#[case("err-prf-iier", proof_err(ProofError::InvalidImportedExitsRoot {
    declared: Digest([0x44; 32]),
    computed: Digest([0x55; 32]),
}))]
#[case("err-ver-vm", ver_err(Pve::VersionMismatch("vm".into())))]
#[case("err-ver-core", ver_err(Pve::Core("core".into())))]
#[case("err-ver-rec", ver_err(Pve::Recursion("rec".into())))]
#[case("err-ver-pl", ver_err(Pve::Plonk("plonk".into())))]
#[case("err-ver-gr", ver_err(Pve::Groth16("groth16".into())))]
#[case("err-ver-pv", ver_err(Pve::InvalidPublicValues))]
#[case("err-ts", err(Cse::TrustedSequencerNotFound(NetworkId::new(6))))]
#[case("err-pr", err(Cse::LastPessimisticRootNotFound(NetworkId::new(5))))]
#[case("err-ie", err(Cse::InternalError("internal".into())))]
#[case("err-se", err(Cse::SettlementError("settlement".into())))]
#[case("err-pce", err(Cse::PreCertificationError("precert".into())))]
#[case("err-ce", err(Cse::CertificationError("cert".into())))]
#[case("err-l1", err(Cse::L1InfoRootNotFound(0xabcd)))]
fn encoding(#[case] name: &'static str, #[case] status: CertificateStatus) {
    // Check for changes in encoding of certificate status.
    // Reordering arms in the status enum causes the storage encoding to change, causing
    // compatibility issues.

    let bytes = Bytes::from(bincode_codec().serialize(&status).unwrap());
    insta::assert_snapshot!(name, bytes);
}
