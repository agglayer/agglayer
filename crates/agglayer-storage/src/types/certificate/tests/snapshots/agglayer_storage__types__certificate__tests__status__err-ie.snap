---
source: crates/agglayer-storage/src/types/certificate/tests/status.rs
expression: bytes
snapshot_kind: text
---
0x00000003000000040000000000000008696e7465726e616c
